### DISCUSSION TITLE
Why reactions should be used sparingly?

### DISCUSSION THREAD

**Discussion by alexkuc** [mobxjs/mobx#3544] [Q&A] [answered]
According to [docs](https://mobx.js.org/reactions.html), reactions should be used sparingly. However, it is not clear why and what's the reason for that. I have checked existing issues and discussions but didn't find a clear-cut answer.

It seems reactions are really convenient when you have internal class flows that depend on number of properties. Let's say a class that is responsible for generating time table, where number of days, starting day, starting hour, ending hour, etc. are all user-configurable. By configuring reactions for each of those properties I can ensur a consistent state of data. However, according to docs, it's not best practise to do so?

**Comment by urugator** [comment, marked as answer]
> I can ensur a consistent state of data.

You can't. All reactions wait until your system is already in a consistent state (that's when action is finished):
```javascript
autorun(() => this.b = this.a + 1)

@action doSomething() {
  this.a + 1;
  console.log(this.b) // inconsistent, not updated yet
}
```
Also there is no guaranteed order:
```javascript
autorun(() => this.b = this.a + 1)
autorun(() => this.c = this.b + 1)
// these can run in whathever order
```

Lastly it (arguably) leads to a bad architecture, don't want to go into lengthy details, just a simple example. Imagine you wanna find out what happens when you click on something:
```javascript
@action onClick() {
  this.a = 5;
}
```
Ok, so `a` changes to 5, simple ... except if you use reactions/autorun, there could be 50 different places in your code base, that do "something", when `this.a` changes, good luck tracking them down. And you can chain these reactions, right...
It's similar to using React with it's one way data flow VS connecting components via events.

I also tried to make some points here, if you're interested:
https://github.com/mobxjs/mobx/issues/1006#issuecomment-426027838

**Reply by alexkuc** [reply]
Thank you @urugator so much!! Also, the #1006 comments were really helpful! Greatly appreciated!!
